lib/reporting.rb

```diff
@@ -0,0 +1,163 @@
+  #
+  def export_modules
+    export('module_status', modules.first.keys, modules.map(&:values))
+  end
+
+  # ----------------------------------------------------------------------------
+
+  def users
+    {
+      # registration scopes
+      registration_complete: User.registration_complete.count,
+      registration_incomplete: User.registration_incomplete.count,
+      reregistered: User.reregistered.count,
+      registered_since_private_beta: User.registered_since_private_beta.count,
+      private_beta_only_registration_incomplete: User.private_beta_only_registration_incomplete.count,
+      private_beta_only_registration_complete: User.private_beta_only_registration_complete.count,
+
+      # registration events
+      registration_events: Ahoy::Event.user_registration.count,
+      private_beta_registration_events: Ahoy::Event.private_beta_registration.count,
+      public_beta_registration_events: Ahoy::Event.public_beta_registration.count,
+
+      # all
+      total: User.all.count,
+
+      # devise
+      locked_out: User.locked_out.count,
+      confirmed: User.confirmed.count,
+      unconfirmed: User.unconfirmed.count,
+
+      # user input
+      user_defined_roles: User.all.collect(&:role_type_other).uniq.count,
+
+      # course engagement
+      started_learning: started_learning,
+      not_started_learning: not_started_learning,
+
+      # Number of distinct users who have at least one note
+      with_notes: get_users_with_notes_count,
+      with_notes_percentage: (get_users_with_notes_count.to_f / User.all.count * 100).round(2),
+
+    }
+  end
+
+  def modules
+    mods.map do |mod|
+      {
+        id: mod.id,
+        name: mod.name,
+        title: mod.title,
+
+        # module_time_to_completion
+        not_started: not_started(mod),
+        started: started(mod),
+        in_progress: in_progress(mod),
+        completed: completed(mod),
+
+        true_false: true_false_count(mod),
+        # module
+        module_start: Ahoy::Event.module_start.where_module(mod.name).count,
+        module_complete: Ahoy::Event.module_complete.where_module(mod.name).count,
+
+        # confidence
+        confidence_check_start: Ahoy::Event.confidence_check_start.where_module(mod.name).count,
+        confidence_check_complete: Ahoy::Event.confidence_check_complete.where_module(mod.name).count,
+
+        # summative
+        start_assessment: Ahoy::Event.summative_assessment_start.where_module(mod.name).count,
+        pass_assessment: Ahoy::Event.summative_assessment_pass(mod.name).count,
+        fail_assessment: Ahoy::Event.summative_assessment_fail(mod.name).count,
+      }
+    end
+  end
+
+private
+
+  # @return [Integer] Number of users who have added at least one note
+  def get_users_with_notes_count
+    User.joins(:notes).distinct.count
+  end
+
```

Comment:
I think we need more than just a count of notes; a note is counted even if it is blank and we add blank notes when they hit "save and return" on a page.  We need to actually scan the body of each note (which is encrypted) to determine if it is blank and only count if is not blank.